### PR TITLE
CFE-3644: Rework acceptance test for `sys.os_version_major`

### PR DESCRIPTION
**Pull request by larsewi**
Rework acceptance test for `sys.os_version_major` to be more dynamic

Ticket: CFE-3644
Changelog: None
Signed-off-by: larsewi <lars.erik.wik@northern.tech>

### CODE REVIEW

**Comment by larsewi** [comment]
@cf-bottom Jenkins with exotics please !

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6526/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6526/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6526/

**Review by craigcomstock — APPROVED** [review]

File: tests/acceptance/01_vars/01_basic/os_version_major.cf [hunk 1]
```diff
@@ -10,105 +16,39 @@ bundle agent test
...
-      "expected" string => "11";
+      # Platforms to test
+      "platforms"
+        slist => { "debian", "ubuntu", "redhat", "rhel", "centos", "fedora",
+                   "aix", "hpux", "suse", "opensuse", "sles", "solaris",
+                   "sunos", "windows", "freebsd", "macos" };
+
+      # Create regex matching os version major classes
+      "regex"
+        string => format("^[%s]+_[0-9]+$", join("|", "platforms"));
```
Comment (vpodzime) [inline tests/acceptance/01_vars/01_basic/os_version_major.cf:27]: I believe `[%s]` should be `(%s)`, i.e. list of substring options not a list of characters.

**Review by vpodzime — COMMENTED** [review]
Maybe below is why Solaris fails? Looks good to me otherwise.

File: tests/acceptance/01_vars/01_basic/os_version_major.cf (hunk 1 above)
Comment (larsewi) [reply tests/acceptance/01_vars/01_basic/os_version_major.cf:28]: I think maybe Solaris is missing a class with just OS name and version major class _(I'll make a ticket)_. I can try this option.

**Comment by larsewi** [comment]
@cf-bottom Jenkins with exotics please !

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6530/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6530/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6530/

**Comment by larsewi** [comment]
@cf-bottom Jenkins with exotics please !

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6531/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6531/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6531/

File: tests/acceptance/01_vars/01_basic/os_version_major.cf [hunk 2]
```diff
@@ -10,105 +16,48 @@ bundle agent test
...
-    hpux_10::
-      "expected" string => "10";
-    hpux_11::
-      "expected" string => "11";
+      # Regex to extract major version number from OS-class
+      # On Solaris/SunOS major version comes second
+      "extract_regex"
+        string => ifelse("solaris|sunos",
+                         "^[a-z]+_[0-9]+_([0-9]+$)",
+                         "^[a-z]+_([0-9]+)_[0-9]+$");
```
Comment (larsewi) [inline tests/acceptance/01_vars/01_basic/os_version_major.cf:34]: I think maybe HP-UX will fail since their os-classes looks something like this `hpux_b_11_23`

**Comment by larsewi** [comment]
@cf-bottom let's try to run Jenkins with exotics one last time. Thanks!

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6533/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6533/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6533/

**Review by olehermanse — CHANGES_REQUESTED** [review]
Still failing on HP-UX, it seems:

```
R: $(regex)
R: $(os_class)
R: /home/jenkins/workspace/testing-pr/label/PACKAGES_ia64_hpux_11.23/cfengine-3.18.0a.00ede4b81/tests/acceptance/./01_vars/01_basic/os_version_major.cf Expected: $(test.expected)
R: /home/jenkins/workspace/testing-pr/label/PACKAGES_ia64_hpux_11.23/cfengine-3.18.0a.00ede4b81/tests/acceptance/./01_vars/01_basic/os_version_major.cf Found: 11
R: /home/jenkins/workspace/testing-pr/label/PACKAGES_ia64_hpux_11.23/cfengine-3.18.0a.00ede4b81/tests/acceptance/./01_vars/01_basic/os_version_major.cf FAIL
```

File: tests/acceptance/01_vars/01_basic/os_version_major.cf [hunk 3]
```diff
@@ -10,105 +16,52 @@ bundle agent test
...
-      "expected" string => "11";
+      # Regex to extract major version number from OS-class
+      # Edge cases:
+      #  - On Solaris/SunOS major version comes second
+      #    E.g. Solaris 11 has class "solaris_5_11"
+      #  - On HP-UX there is an additional model letter
+      #    E.g. HP-UX 11 has class "hpux_b_11_23"
+      "extract_regex"
+        string => ifelse("solaris|sunos", "^[a-z]+_[0-9]+_([0-9]+$)",
+                         "hpux",          "^[a-z]+_[a-z]_([0-9]+)_[0-9]+$",
```
Comment (olehermanse) [inline tests/acceptance/01_vars/01_basic/os_version_major.cf:37]: @larsewi I am assuming that `hpux` is actually defined - if not we have a problem.

Then, according to this test `hpux_11` should also be defined:

https://github.com/cfengine/core/blob/86079af65b1f6a7b4069d0a07837e4fcb712a3cc/tests/acceptance/02_classes/01_basic/hp_ux_major.cf

Maybe your regex is too generic and matching other things? What if it was just:

```
hpux_([0-9]+)$
```

**Comment by larsewi** [comment]
@cf-bottom Jenkins with exotics please !

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6542/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6542/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6542/

**Comment by larsewi** [comment]
@cf-bottom Jenkins with exotics please !

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6545/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6545/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6545/

File: tests/acceptance/01_vars/01_basic/os_version_major.cf (hunk 3 above)
Comment (larsewi) [reply tests/acceptance/01_vars/01_basic/os_version_major.cf:37]: Just checked, both `hpux` and `hpux_11` are defined. No need to worry. On the other hand `hpux_b_11_23` was not defined, but `hp_ux_b_11_23` was defined :wink:

**Review by vpodzime — DISMISSED** [review]
Looks promising to me.

**Comment by larsewi** [comment]
@cf-bottom Jenkins with exotics please !

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6546/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6546/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6546/

**Comment by larsewi** [comment]
@cf-bottom Jenkins with exotics please !

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6548/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6548/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6548/

**Review by olehermanse — APPROVED** [review]
Looks good to me.

**Review by vpodzime — APPROVED** [review]
ACK

**Comment by larsewi** [comment]
@cf-bottom Jenkins with exotics please !

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6571/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6571/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6571/

**Comment by larsewi** [comment]
@cf-bottom Jenkins with exotics please !

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6578/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6578/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6578/
